2023/09/marvinhsu/src/lib.rs

```diff
@@ -0,0 +1,94 @@
+use anyhow::{Context, Result};
+use itertools::Itertools;
+
+pub fn part1(input: &str) -> Result<isize> {
+    input
+        .lines()
+        .map(|line| {
+            line.split_ascii_whitespace()
+                .map(|s| s.parse().context("Failed to parse input"))
+                .collect::<Result<Vec<_>>>()
+        })
+        .process_results(|iter| iter.map(|layer| get_next_number(layer.as_slice())).sum())
+}
+
+pub fn part2(input: &str) -> Result<isize> {
+    input
+        .lines()
+        .map(|line| {
+            line.split_ascii_whitespace()
+                .map(|s| s.parse().context("Failed to parse input"))
+                .collect::<Result<Vec<_>>>()
+        })
+        .process_results(|iter| {
+            iter.map(|layer| get_previous_number(layer.as_slice()))
+                .sum()
+        })
+}
+
+fn get_sub_layer(layer: &[isize]) -> Option<Vec<isize>> {
+    match layer {
+        [.., 0] => None,
+        _ => Some(
+            layer
+                .iter()
+                .zip(layer.iter().skip(1))
+                .map(|(a, b)| b - a)
+                .collect_vec(),
+        ),
+    }
+}
+
+fn get_next_number(layer: &[isize]) -> isize {
+    let last = layer[layer.len() - 1];
+    match get_sub_layer(layer) {
+        Some(sub_layer) => last + get_next_number(sub_layer.as_slice()),
+        None => last,
+    }
+}
+
+fn get_previous_number(layer: &[isize]) -> isize {
+    let first = layer[0];
+    match get_sub_layer(layer) {
+        Some(sub_layer) => first - get_previous_number(sub_layer.as_slice()),
+        None => first,
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::vec;
```

Comment:
This is not necessary. vec is in prelude https://doc.rust-lang.org/std/prelude/index.html